### PR TITLE
defendguin: Now loads from and saves in Haiku's settings directory.

### PR DESCRIPTION
**Pull request by R4H33M**
Tested on x64

### CODE REVIEW

File: games-arcade/defendguin/patches/defendguin-0.0.12.patchset [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
++#endif
+ 
+       fi = fopen(fname, "r");
+ 
+@@ -5948,7 +5958,13 @@ void save_options(void)
+   
+   if (getenv("HOME") != NULL)
+     {
++#ifdef __HAIKU__
++      BPath* s_path = new BPath();
```
Comment (korli) [inline games-arcade/defendguin/patches/defendguin-0.0.12.patchset]: This is C++, please use the C API
```
      char path[PATH_MAX];
      dev_t bootdev = dev_for_path("/boot");
      if (bootdev < 0)
        bootdev = 0;
      if (find_directory(B_USER_SETTINGS_DIRECTORY, bootdev, false,
          path, sizeof(path) - 1) == B_OK) {`

**Comment by R4H33M** [comment]
Ready for merge?

File: games-arcade/defendguin/defendguin-0.0.12.recipe [hunk 2]
```diff
@@ -8,6 +8,7 @@ LICENSE="GNU GPL v2"
 REVISION="3"
```
Comment (korli) [inline games-arcade/defendguin/defendguin-0.0.12.recipe]: please bump the revision.

File: games-arcade/defendguin/patches/defendguin-0.0.12.patchset [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   if (getenv("HOME") != NULL)
+     {
++#ifdef __HAIKU__
++      char path[PATH_MAX];
++      dev_t bootdev = dev_for_path("/boot");
++      if (bootdev < 0)
++      	bootdev = 0;
++      if (find_directory(B_USER_SETTINGS_DIRECTORY, bootdev, false,
++        path, sizeof(path) - 1) == B_OK) {
++          snprintf(fname, sizeof(fname), "%s/.defendguinrc", path);
```
Comment (korli) [inline games-arcade/defendguin/patches/defendguin-0.0.12.patchset]: please use defendguinrc without dot prefix.

**Comment by Begasus** [comment]
@korli you are the master here :) ready for merge?

**Comment by korli** [comment]
Build warnings:
cc -Wall -I/packages/libsdl-1.2.15-15/.self/develop/headers/SDL -D_GNU_SOURCE=1 -DDATA_PREFIX=\"/packages/defendguin-0.0.12-4/.self/data/defendguin/\" -D__SOUND -DJOY_YES        src/defendguin.c -c -o obj/defendguin.o
src/defendguin.c: In function 'load_options':
src/defendguin.c:5892:23: warning: implicit declaration of function 'dev_for_path' [-Wimplicit-function-declaration]
      dev_t bootdev = dev_for_path("/boot");
                      ^~~~~~~~~~~~
src/defendguin.c:5897:43: warning: '%s' directive output may be truncated writing up to 1023 bytes into a region of size 512 [-Wformat-truncation=]
          snprintf(fname, sizeof(fname), "%s/defendguinrc", path);
                                          ^~                ~~~~
src/defendguin.c:5897:11: note: 'snprintf' output between 14 and 1037 bytes into a destination of size 512
          snprintf(fname, sizeof(fname), "%s/defendguinrc", path);
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/defendguin.c: In function 'save_options':
src/defendguin.c:5972:43: warning: '%s' directive output may be truncated writing up to 1023 bytes into a region of size 512 [-Wformat-truncation=]
          snprintf(fname, sizeof(fname), "%s/defendguinrc", path);
                                          ^~                ~~~~
src/defendguin.c:5972:11: note: 'snprintf' output between 14 and 1037 bytes into a destination of size 512
          snprintf(fname, sizeof(fname), "%s/defendguinrc", path);
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by korli** [comment]
https://build.haiku-os.org/buildmaster/master/x86_64/logviewer.html?buildruns/3984/builds/15025.log

**Comment by billkendrick** [comment]
I've taken this and worked into upstream. I actually created a new function to handle figuring out the config filename (to reduce duplicated code), and if the Haiku stuff fails, it'll fallback to using `$HOME/.defendguinrc` (and if `$HOME` isn't set for some reason, it just creates "`defendguinrc`" in the working directory).

Also, for the "`-lbe`" change to `Makefile`, I've handed that via a new "`EXTRA_LIBS`" Makefile variable.

Look for this in Defendguin 0.0.13 (along with a ton of art & sfx changes, a few bugfixes, and a new default 800x480 window size). Thanks!

**Comment by Begasus** [comment]
@billkendrick Thanks for keeping track here! :) Could you point to the new source so we could check it out before you release a 0.0.13?

**Comment by billkendrick** [comment]
> @billkendrick Thanks for keeping track here! :)

Scott was kind enough to point me to this :)

> Could you point to the new source so we could check it out before you release a 0.0.13?

It's not in a source code repository, so I just rolled a tarball of what I was working on yesterday. It's not ready for release (a lot of sound effects are silent, waiting to be replaced), but you can hopefully build it right for Haiku. Let me know if it needs any adjustment.

ftp://billsgames.com/unix/x/defendguin/src/defendguin-0.0.13-pre-2020.01.19.tar.gz

**Pull request merged**